### PR TITLE
#169817546 Requesting one way trip

### PR DESCRIPTION
**Pull request by gadishimwe**
### What does this PR do?
Add one way trip request feature

### Description of Task to be completed?
- Add one way trip endpoint
- Add validations
- Add tests for one way trips

### How should this be manually tested?
```
$ git clone https://github.com/andela/the_spinners-backend.git
$ cd the_spinners-backend
$ git checkout  ft-one-way-trip-request-169817546
```
then 
`$ npm run test`

or you can use Postman to send the request which contains departure, destination, travelDate, travelReasons and accommodation as request body.

### What are the relevant pivotal tracker stories?

[start #169817546]
### Screenshots
Screen Shot 2019-12-26 at 16 57 03
Screen Shot 2019-12-26 at 16 56 06

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,3 +1,12 @@
 {
-  "presets": ["@babel/preset-env"]
-}
+  "presets": [
+    "@babel/preset-env"
+  ],
+  "env": {
```
Comment (adesege) [inline .babelrc:5]: Why this extra config?

File: src/services/tripServieces.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import database from '../models';
```
Comment (adesege) [inline src/services/tripServieces.js]: This is not a database but an encapsulation of all the models.

**Review by adesege — CHANGES_REQUESTED** [review]

File: .babelrc (hunk 1 above)
Comment (gadishimwe) [reply .babelrc:5]: It's for getting 100% coverage on async functions, I added Istanbul to achieve that.
Screen Shot 2019-12-09 at 15 23 18
before
Screen Shot 2019-12-09 at 15 44 49
After

File: src/services/tripServieces.js (hunk 2 above)
Comment (gadishimwe) [reply src/services/tripServieces.js]: I'm gonna rename it

File: package.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "start": "node production/index.js",
     "dev": "nodemon --exec babel-node src/index.js",
     "coverage": "nyc report --reporter=text-lcov | coveralls",
-    "test": "nyc --reporter=html --reporter=text mocha  --require  @babel/register  src/tests/*.js --exit",
+    "test": "export NODE_ENV=test && sequelize db:migrate:undo:all  && sequelize db:migrate && nyc --reporter=html --reporter=text mocha  --require  @babel/register --require @babel/polyfill  src/tests/*.js --exit",
```
Comment (victor-abz) [inline package.json]: Here remove the `export NODE_ENV=test && sequelize db:migrate:undo:all  && sequelize db:migrate &&`
 add:
```suggestion
    "pretest": "export NODE_ENV=test && sequelize db:migrate:undo:all  && sequelize db:migrate",
    "test": "sequelize db:migrate && nyc --reporter=html --reporter=text mocha  --require  @babel/register --require @babel/polyfill  src/tests/*.js --exit",
```

File: src/tests/trips.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+describe('Test one way trip:', () => {
+  it('Should return status code of 400 on invalid input into departure', (done) => {
+    const newTrip = {
+      departure: '',
+      destination: 'London',
+      travelDate: '2019-12-21',
+      travelReasons: 'Visiting',
+      accommodation: 'Some hotel'
+    };
```
Comment (victor-abz) [inline src/tests/trips.js]: User faker.js to generate data for you

File: src/index.js [hunk 5]
```diff
@@ -1,9 +1,17 @@
...
 import swaggerDefinition from './swagger';
+import trips from './routes/trips';
+
 
 const app = express();
+app.use(morgan('dev'));
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+
+app.use('/', trips);
```
Comment (victor-abz) [inline src/index.js]: you can add the `api` to make prefix for your routes

**Review by victor-abz — CHANGES_REQUESTED** [review]
- It shouldn't be able to allow past dates for requests..

- it shouldn't return a null user.... You should get the user who is requesting for trip at least

Screen Shot 2019-12-11 at 15 19 16

- Validation for dates should be strong... and not return incorrect values when entered in a wrong format.

**Comment by gadishimwe** [comment]
> * It shouldn't be able to allow past dates for requests..
> * it shouldn't return a null user.... You should get the user who is requesting for trip at least
> 
> Screen Shot 2019-12-11 at 15 19 16
> 
> * Validation for dates should be strong... and not return incorrect values when entered in a wrong format.

I can't remove userId, it will not be null if token guy merge

**Review by victor-abz — APPROVED** [review]

File: src/controllers/trip.controller.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import TripService from '../services/trip.service';
+import Response from '../services/response.service';
+
+
+/**
+ *
+ *
+ * @class Trips
```
Comment (victor-abz) [inline src/controllers/trip.controller.js]: This is not class trip. Update comments

File: src/tests/trip/trips.test.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      .post('/api/one-way-trips')
+      .set('Authorization', loggedInToken)
+      .send({ ...newTrip, departure: '' })
+      .end((err, res) => {
+        expect(res).to.have.status(400);
+        expect(res.body).to.have.property('message');
+        done();
+      });
+  });
+  it('Should return status code of 400 on invalid input into destination', (done) => {
```
Comment (victor-abz) [inline src/tests/trip/trips.test.js:52]: Positive(Success ) Tests should come first then Negative test come after as agreed .

**Review by victor-abz — CHANGES_REQUESTED** [review]
- Make sure your feature is not duplicating data Multiple requests with the same destination, dates,...

- The error message should be more descriptive(e.g: dates shouldn't be in the past  and if invalid provides an example like dd/mm/yyyy...)

Screen Shot 2019-12-23 at 16 14 23

- I think it is not necessary to return all fields in the database. Especially (created at and updated at)
Screen Shot 2019-12-23 at 16 16 35

**Review by higustave-ops — APPROVED** [review]
LGTM

**Review by victor-abz — APPROVED** [review]
LGTM

**Comment by katunold** [comment]
@gadishimwe kindly squash the commits into one commit

**Review by katunold — COMMENTED** [review]
@gadishimwe kindly reword your commit message to look a bit more organised with a clear header, body and footer section.

File: .babelrc [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
   "env": {
     "test": {
       "plugins": [
+        "istanbul",
         "@babel/plugin-transform-runtime"
       ]
     }
   }
-  
 }
```
Comment (katunold) [inline .babelrc]: Kindly add a new line at the end of this file

File: src/app.js [hunk 9]
```diff
@@ -10,5 +11,6 @@ app.use(express.urlencoded({ extended: false }));
 app.use(express.json());
 
 app.use('/api/auth', authRoute);
+app.use('/api/', tripRoute);
```
Comment (katunold) [inline src/app.js]: How about you create a single module that handles all routing and it's that module you import and use here 🤔

File: src/swagger/trip.swagger.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+ *              - travelReasons
+ *              - accommodation
+ *     responses:
+ *       '201':
+ *         description: Trip created successfully
+ *       '400':
+ *         description: invalid inputs.
+ *       '401':
+ *         description: No Token supplied
+ */
```
Comment (katunold) [inline src/swagger/trip.swagger.js:70]: Kindly do away with commented out stuff you are not using unless if you expect to use it in the future

**Review by katunold — COMMENTED** [review]
@gadishimwe, please kindly consider the feedback I have left behind and then all will be good. Otherwise well done

File: src/swagger/trip.swagger.js (hunk 10 above)
Comment (gadishimwe) [reply src/swagger/trip.swagger.js:70]: @katunold We're using this to document our API. Give me more clarifications if I'm not getting it well

**Pull request merged**